plugins/reindex/src/main/java/org/elasticsearch/plugin/reindex/AbstractAsyncBulkIndexByScrollAction.java

```diff
@@ -0,0 +1,228 @@
+            bulkRequest.add(index);
+        }
+
+        return bulkRequest;
+    }
+
+    /**
+     * Copies the metadata from a hit to the index request.
+     */
+    protected void copyMetadata(IndexRequest index, SearchHit doc) {
+        index.parent(fieldValue(doc, "_parent"));
+        copyRouting(index, doc);
+        // Comes back as a Long but needs to be a string
+        Long timestamp = fieldValue(doc, "_timestamp");
+        if (timestamp != null) {
+            index.timestamp(timestamp.toString());
+        }
+        index.ttl(fieldValue(doc, "_ttl"));
+    }
+
+    /**
+     * Part of copyMetadata but called out individual for easy overwriting.
+     */
+    protected void copyRouting(IndexRequest index, SearchHit doc) {
+        index.routing(fieldValue(doc, "_routing"));
+    }
+
+    protected <T> T fieldValue(SearchHit doc, String fieldName) {
+        SearchHitField field = doc.field(fieldName);
+        return field == null ? null : field.value();
+    }
+
+    /**
+     * Apply a script to the request.
+     *
+     * @return is this request still ok to apply (true) or is it a noop (false)
+     */
+    @SuppressWarnings("unchecked")
+    protected boolean applyScript(IndexRequest index, SearchHit doc, ExecutableScript script, final Map<String, Object> ctx) {
+        if (script == null) {
+            return true;
+        }
+        ctx.put("_index", doc.index());
+        ctx.put("_type", doc.type());
+        ctx.put("_id", doc.id());
+        Long oldVersion = doc.getVersion();
+        ctx.put("_version", oldVersion);
+        String oldParent = fieldValue(doc, "_parent");
+        ctx.put("_parent", oldParent);
+        String oldRouting = fieldValue(doc, "_routing");
+        ctx.put("_routing", oldRouting);
+        Long oldTimestamp = fieldValue(doc, "_timestamp");
+        ctx.put("_timestamp", oldTimestamp);
+        Long oldTtl = fieldValue(doc, "_ttl");
+        ctx.put("_ttl", oldTtl);
+        ctx.put("_source", index.sourceAsMap());
+        ctx.put("op", "update");
+        script.setNextVar("ctx", ctx);
+        script.run();
+        Map<String, Object> resultCtx = (Map<String, Object>) script.unwrap(ctx);
+        String newOp = (String) resultCtx.remove("op");
+        if (newOp == null) {
+            throw new IllegalArgumentException("Script cleared op!");
+        }
+        if ("noop".equals(newOp)) {
+            noops.incrementAndGet();
+            return false;
+        }
+        if ("update".equals(newOp) == false) {
+            throw new IllegalArgumentException("Invalid op [" + newOp + ']');
+        }
+
+        /*
+         * It'd be lovely to only set the source if we know its been
+         * modified but it isn't worth keeping two copies of it
+         * around just to check!
+         */
+        index.source((Map<String, Object>) resultCtx.remove("_source"));
+
+        Object newValue = ctx.remove("_index");
+        if (false == doc.index().equals(newValue)) {
```

Comment:
I like this way better actually. If we're not going to do `!` for checking false stuff I think putting the false on the left hand side is more readable than the right. So I'll try and change the other pattern.